quotecache/quotecache.go

```diff
@@ -10,43 +10,50 @@ import (
 	"strconv"
 	"strings"
 	"time"
+
+	"github.com/distributeddesigns/currency"
 )
 
 // Quote : Stored response from the quoteserver
 type Quote struct {
 	UserID    string
 	Stock     string
-	Price     float64
+	Price     currency.Currency
 	Timestamp time.Time
 	Cryptokey string
 }
 
 // IsExpired : True if the quotes timestamp is older than its validity window
 func (q Quote) IsExpired() bool {
-	expiry := q.Timestamp.Add(time.Second * 45)
+	expiry := q.Timestamp.Add(time.Second * 60)
 	return time.Now().After(expiry)
 }
 
-// Global to store cached responses.
-// Maps stock name -> Quote
-var cache = make(map[string]Quote)
+var QuoteCache = make(map[string]map[string]Quote)
 
 // GetQuote : Gets the current value of the stock, hitting the local cache if it can.
 func GetQuote(userID, stock string) (Quote, error) {
 	// check if the value is in cache
-	quote, found := cache[stock]
 
-	if !found || quote.IsExpired() {
+	var userQuote Quote
+	userMap := QuoteCache[stock]
+	userQuote, found := userMap[userID]
+	if found && !userQuote.IsExpired() {
+		//Get it from the cache
+		return userQuote, nil
+	}
+	//Failed to get from cache, go do it outselves.
+
+	if !found || userQuote.IsExpired() {
```

Comment:
This is the negation of the condition at L41 so we can run this code unconditionally (or wrap it in an `else` for extra safety.